File: src/core/responder/responder.ts
let firstResponder: Responder | null = null;

// The chain-top responder — Cocoa's NSWindow in the FR fallback. Set to
// the `Application` at boot (see `setChainTop`). On an incoming refusal
// or a `null` target, FR falls here rather than to `null`, exactly as
// Cocoa bottoms out at the window/app.
// TODO(per-window FR): when responder chains go per-window, the chain
// top becomes that window's responder and this module-global splits.
let chainTop: Responder | null = null;

/** Current first responder, or null. Read-only outside this module. */
export function currentFirstResponder(): Responder | null {
  return firstResponder;
}

/** Wire the chain-top fallback responder (the `Application`). Called once
 *  at boot. TODO(per-window FR): becomes per-window. */
export function setChainTop(r: Responder): void {
  chainTop = r;
}

/** The chain-top responder (the `Application`), or null pre-boot. The
 *  responder chain bottoms out here — a root controller's
 *  `nextResponder` is this. TODO(per-window FR): per-window. */
export function chainTopResponder(): Responder | null {
  return chainTop;
}

/**
 * Install a new first responder after DOM focus has shifted. Internal —
 * `syncFirstResponderFromFocus` uses it to record the FR the browser
 * already moved focus to. Prefer `makeFirstResponder`.
 */
export function _setFirstResponder(r: Responder | null): void {
  firstResponder = r;
}

/**
 * Cocoa `NSWindow.makeFirstResponder(_:)` — the single orchestrator for
 * changing the first responder. Verified against Apple's documented
 * contract:
 *
 * 1. `responder` already FR → `true`, send nothing.
 * 2. Send `resignFirstResponder(responder)` to the current FR. **Refusal
 *    aborts**: FR unchanged, return `false` (the caller restores DOM
 *    focus — that's a View concern, not this module's).
 * 3. Current FR resigned. `responder == null` → FR = chain-top, `true`.
 * 4. Send `becomeFirstResponder()` to `responder`. **Refusal does NOT
 *    abort and does NOT return false** — FR = chain-top, return `true`.
 * 5. Accepted → FR = `responder`, `true`.
 *
 * Does NOT pre-gate on `canBecomeFirstResponder` — per Apple that's the
 * caller's check. The outgoing/incoming asymmetry is deliberate: only an
 * outgoing resign-refusal aborts.
 */
export function makeFirstResponder(responder: Responder | null): boolean {
  if (responder === firstResponder) return true;            // 1
  const current = firstResponder;
  if (current && !current.resignFirstResponder(responder)) return false; // 2
  firstResponder = null;
  if (responder == null) {                                  // 3
    firstResponder = chainTop;
    return true;
  }
  if (responder.becomeFirstResponder()) {                   // 5
    firstResponder = responder;
    return true;
  }
  firstResponder = chainTop;                                // 4
  return true;
}

/**
 * Abstract base for anything that participates in the responder chain.
 * Cocoa NSResponder / UIResponder analog.
 *
 * Pure chain mechanics: first-responder state, event routing, and
 * target-action lookup. Nothing here knows about views, DOM, or
 * lifecycle — those live on `Controller`, which extends this.
 *
 * @remarks
 * A subclass must declare:
 *
 * - `get nextResponder()`: the next link in the chain (null at the root).
 *
 * It may override:
 *
 * - `canBecomeFirstResponder` / `canResignFirstResponder` (default: true).
 * - `becomeFirstResponder` / `resignFirstResponder` — pure protocol:
 *   accept/refuse + the notification + setup. The handshake + global FR
 *   live in `makeFirstResponder` (no `did*` hooks; Cocoa has none).
 * - Any of `keyDown` / `keyUp` / `cancelOperation` / `insertNewline` / `complete`.
 * - `performAction` for custom dispatch (rare).
 */
export abstract class Responder {
  /** Next link in the responder chain. Cocoa: -[NSResponder nextResponder].
   * Subclasses provide this; `Controller` derives it from View containment. */
  abstract get nextResponder(): Responder | null;

  /** Cocoa NSResponder predicates. Override to gate transitions declaratively;
   * the side-effecting `becomeFirstResponder`/`resignFirstResponder` consult
   * these before doing anything. */
  get canBecomeFirstResponder(): boolean { return true; }
  get canResignFirstResponder(): boolean { return true; }

  get isFirstResponder(): boolean { return firstResponder === this; }

  /**
   * Pure protocol (Cocoa `NSResponder.becomeFirstResponder()`): the
   * accept-or-refuse decision *and* the "about to become FR"
   * notification. Return `false` to refuse. Do setup here (focus,
   * etc.). Does NOT touch the global FR or run the handshake — that's
   * `makeFirstResponder`'s job. There is no separate `did*` hook;
   * `become`/`resign` ARE the notifications (Cocoa has none).
   */
  becomeFirstResponder(): boolean {
    return this.canBecomeFirstResponder;
  }

  /**
   * Pure protocol (Cocoa `NSResponder.resignFirstResponder()`). Return
   * `false` to refuse — `makeFirstResponder` then aborts the transition
   * and the outgoing responder stays FR. `next` is the proposed
   * incoming responder (null when focus is going nowhere); subclasses
   * inspect it to delegate, validate, or prompt before allowing.
   */
  resignFirstResponder(_next: Responder | null = null): boolean {
    return this.canResignFirstResponder;
  }

  // ---- event chain ----
  //
  // Default behavior matches Cocoa's NSResponder: keyDown routes well-known
  // keys to named methods (cancelOperation/insertNewline/complete); other
  // keys, and the named methods themselves, bubble up via nextResponder.

  keyDown(event: KeyboardEvent): void {
    switch (event.key) {
      case "Escape": return this.cancelOperation(event);
      case "Enter":  return this.insertNewline(event);
      case "Tab":    return this.complete(event);
      default:       this.nextResponder?.keyDown(event);
    }
  }
  keyUp(event: KeyboardEvent): void { this.nextResponder?.keyUp(event); }
  cancelOperation(event: KeyboardEvent): void { this.nextResponder?.cancelOperation(event); }
  insertNewline(event: KeyboardEvent): void { this.nextResponder?.insertNewline(event); }
  complete(event: KeyboardEvent): void { this.nextResponder?.complete(event); }

  /** Walk the responder chain looking for a method with this name. Used by
   * target-action: `<button data-action="save">` starts at
   * `firstResponderFor(buttonElement)`, then walks up until something
   * handles it. */
  performAction(name: string, sender: unknown, event?: Event): boolean {
    let node: Responder | null = this;
    while (node) {
      const fn = (node as any)[name];
      if (typeof fn === "function") {
        fn.call(node, sender, event);
        return true;
      }
      node = node.nextResponder;
    }
    return false;
  }
}

File: opal/lib/swill/responder.rb
# frozen_string_literal: true

module Swill
  # The first-responder singleton and the one orchestrator that changes it.
  # Holds module-level state (Opal is single-threaded, so no per-thread
  # concern) rather than per-class, so that calling through any Responder
  # subclass operates on the same first responder.
  #
  # TODO(per-window FR): when responder chains become per-window, `chain_top`
  # and `current` move onto the window.
  module FirstResponder
    @current = nil
    @chain_top = nil

    class << self
      # The current first responder, or nil.
      attr_reader :current

      # The chain-top responder (the Application), set once at boot. A root
      # controller's next_responder bottoms out here; on an incoming refusal or
      # a nil target the first responder falls here rather than to nil —
      # Cocoa's NSWindow/app fallback.
      attr_accessor :chain_top

      # Record the first responder after DOM focus has already moved. Internal;
      # the focus adapter uses it. Prefer `make`.
      def install(responder)
        @current = responder
      end

      # Cocoa NSWindow#makeFirstResponder — the single orchestrator. Verified
      # against Apple's documented contract:
      #
      #   1. already the FR             -> true, do nothing
      #   2. ask current FR to resign   -> refusal ABORTS (FR unchanged, false)
      #   3. resigned, target is nil    -> FR = chain_top, true
      #   4. ask target to become       -> refusal does NOT abort: FR = chain_top, true
      #   5. accepted                   -> FR = target, true
      #
      # The outgoing/incoming asymmetry is deliberate: only an outgoing
      # resign-refusal aborts. We do not pre-gate on can_become_first_responder?
      # — per Apple that is the caller's check.
      def make(responder)
        return true if responder.equal?(@current) # 1

        current = @current
        return false if current && !current.resign_first_responder(responder) # 2

        # Clear before `become` so a focus event triggered during become sees no
        # outgoing FR and does not re-run the resign handshake.
        @current = nil

        if responder.nil? # 3
          @current = @chain_top
          return true
        end

        if responder.become_first_responder # 5
          @current = responder
          return true
        end

        @current = @chain_top # 4
        true
      end
    end
  end

  # Base for anything in the responder chain (Cocoa NSResponder). Pure chain
  # mechanics: first-responder protocol, key-event routing, and target-action
  # lookup. Nothing here touches the DOM — View and Controller add that.
  class Responder
    # Next link in the chain; nil at the root. Subclasses provide it.
    def next_responder
      nil
    end

    # Override to gate transitions declaratively; become/resign consult these.
    def can_become_first_responder?
      true
    end

    def can_resign_first_responder?
      true
    end

    def first_responder?
      FirstResponder.current.equal?(self)
    end

    # Request that +responder+ become the first responder. Delegates to the
    # orchestrator so any responder can initiate (a controller asking in
    # after_load, say). Returns whether the FR now reflects the request.
    def make_first_responder(responder)
      FirstResponder.make(responder)
    end

    # Pure protocol (Cocoa becomeFirstResponder): accept-or-refuse plus the
    # "about to become" setup. Return false to refuse. Does not touch the
    # global FR — that is the orchestrator's job; there is no separate did* hook.
    def become_first_responder
      can_become_first_responder?
    end

    # Pure protocol (Cocoa resignFirstResponder). Return false to refuse, which
    # aborts the transition and keeps this responder first. +next_responder+ is
    # the proposed incoming responder (nil when focus is going nowhere).
    def resign_first_responder(_next_responder = nil)
      can_resign_first_responder?
    end

    # ---- key-event chain ----
    #
    # keyDown routes well-known keys to named methods; everything else, and the
    # named methods themselves, bubble up through next_responder (Cocoa).

    def key_down(event)
      case event && event.key
      when "Escape" then cancel_operation(event)
      when "Enter" then insert_newline(event)
      when "Tab" then complete(event)
      else next_responder&.key_down(event)
      end
    end

    def key_up(event)
      next_responder&.key_up(event)
    end

    def cancel_operation(event)
      next_responder&.cancel_operation(event)
    end

    def insert_newline(event)
      next_responder&.insert_newline(event)
    end

    def complete(event)
      next_responder&.complete(event)
    end

    # Walk the chain for the first responder answering +name+. Target-action:
    # `<button data-action="save">` starts at the responder for the button and
    # walks up until something handles it.
    def perform_action(name, sender = nil, event = nil)
      responder = self
      while responder
        if responder.respond_to?(name)
          responder.public_send(name, sender, event)
          return true
        end

        responder = responder.next_responder
      end

      false
    end
  end
end
